README.md
# witchcraft-plugin

A Clojure plugin for Bukkit-based minecraft servers (Paper, Spigot, Glowstone)

It provides

- an embedded nREPL server for interactive creative coding
- Clojure dependency management through `deps.edn`
- loading of Clojure-based server modifications at boot

## Installation

Download the plugin for your server type and version from the
[releases](https://github.com/lambdaisland/witchcraft-plugin/releases) page, and
drop it in your server's `plugins/` directory.

## Usage

The first time you start the server you should see something like this:

```
[09:33:14 INFO]: [Witchcraft] Enabling Witchcraft v0.0.7
[09:33:14 INFO]: [Witchcraft] Setting PluginClassloader as Context classloader: PluginClassLoader{plugin=Witchcraft v0.0.7, pluginEnabled=true, url=plugins/witchcraft-plugin-0.0.7-for-paper-1.17-shaded.jar}
[09:33:26 INFO]: [Witchcraft] No plugins/witchcraft.edn found, creating default.
[09:33:26 INFO]: [Witchcraft] No deps.edn found, creating default.
[09:33:29 INFO]: [Witchcraft] init: (require (quote lambdaisland.witchcraft))
[09:33:30 INFO]: [Witchcraft] init: (lambdaisland.witchcraft/init-xmaterial!)
[09:33:36 INFO]: nREPL server started on port 25555 on host localhost - nrepl://localhost:25555
```

Now you can connect with your nREPL-capable editor to port 25555 and start
manipulating the game. If you get any errors at this stage please file [an
issue](https://github.com/lambdaisland/witchcraft-plugin/issues).

If you are new to Clojure and are unsure which development environment to pick
then we recommend [Calva](https://calva.io/) for Visual Studio Code. For a
fairly complete list of Clojure-capable editors see this [overview by
practical.li](https://practical.li/clojure/clojure-editors/).

## Configuration

There are two files that are used to configure the plugin,
`plugins/witchcraft.edn`, and `deps.edn`. Both are in the [EDN
format](https://github.com/edn-format/edn) (Extensible Data Notation) commonly
used in the Clojure world.

`plugins/witchcraft.edn` is the configuration for the plugin itself. Here you
can set things like the nREPL port, or configure code to load when the server
starts.

`deps.edn` is Clojure's way of declaring dependencies, similar to a `pom.xml` or
`package.json`.

The first time you run the server it will create a `deps.edn` and
`plugins/witchcraft.edn`. It will also start an nREPL server, and invoke any
`:init` commands you have in your `witchcraft.edn`.

### witchcraft.edn

```clojure
{:nrepl
 {:port 25555
  :middleware [refactor-nrepl.middleware/wrap-refactor
               cider.nrepl/cider-middleware]}
 :init [(require 'lambdaisland.witchcraft)
        (lambdaisland.witchcraft/init-xmaterial!)]
 :deps true}
```

- `:nrepl` are options passed to `nrepl.cmdline/dispatch-commands`, if you don't
  want to run an nREPL server, use `:nrepl false`
- `:deps` can be `true` (load dependencies from `deps.edn`), `false` (don't), or
  a map with options for tools.deps, e.g. `:deps {:aliases [:my-alias]}`
- `:init` is a sequence of Clojure forms that are evaluated after starting the
  server. You can use this to load additional Clojure modifications when booting
  the server
  
### deps.edn

The plugin JAR only bundles Clojure itself (the compiler and core library),
nREPL (for interaction), and clojure.tools.deps (for `deps.edn` support). Any
additional dependencies (e.g. nREPL middleware,
[Witchcraft](https://github.com/lambdaisland/witchcraft),
[clj-minecraft](https://github.com/CmdrDats/clj-minecraft)) can be loaded by
adding them to `deps.edn`.

Through `deps.edn` you can pull in dependencies from Maven/Clojars, or directly
from Github. See the [Clojure Deps and CLI Guide](https://clojure.org/guides/deps_and_cli) 
for how to set up `deps.edn`.

When no `deps.edn` is present the plugin will create one on first use, which
loads the [witchcraft library](https://github.com/lambdaisland/witchcraft), as
well as the CIDER and refactor-nrepl middlewares for nREPL, which will look
something like this:

```clojure
{:deps
 {com.lambdaisland/witchcraft
  {:git/url "https://github.com/lambdaisland/witchcraft"
   :git/sha "..."}
  refactor-nrepl/refactor-nrepl {:mvn/version "2.5.1"}
  cider/cider-nrepl {:mvn/version "0.26.0"}}}
```

You can use [lambdaisland/classpath](https://github.com/lambdaisland/classpath)
to reload your `deps.edn` without having to restart the server. (see
`lambdaisland.classpath/update-classpath!` or
`lambdaisland.classpath.watch-deps/start!`).

## Build

To build the plugin yourself you need a recent version of Clojure CLI, see the
[Clojure getting started](https://clojure.org/guides/getting_started) guide. You
also need Java 16 (earlier versions from Java 12 onwards might work too).

```
$ clojure --version
Clojure CLI version 1.10.3.943

$ java -version
openjdk version "16.0.1" 2021-04-20
```

Then use the build-plugin task, you need to tell it which server you will be
using, and the API version. 

```
clojure -T:build-plugin build :server glowstone :api-version 1.12
clojure -T:build-plugin build :server paper :api-version 1.17 
clojure -T:build-plugin build :server spigot :api-version 1.17 
```

These will create two files each 

- `target/witchcraft-plugin-<version>-for-<server>-<api-version>.jar`
- `target/witchcraft-plugin-<version>-for-<server>-<api-version>-shaded.jar`

It is recommended to use the shaded version. For Glowstone you must use the
shaded version. (Naether pulls in an old plexus-utils, which conflicts with
tools.deps.alpha).

Copy the plugin jar to your server's `plugins` directory

```
$ cp target/witchcraft-plugin-0.0.7-for-paper-1.17-shaded.jar ~/PaperMC/plugins/
```

<!-- opencollective -->
## Lambda Island Open Source

<img align="left" src="https://github.com/lambdaisland/open-source/raw/master/artwork/lighthouse_readme.png">

&nbsp;

witchcraft-plugin is part of a growing collection of quality Clojure libraries created and maintained
by the fine folks at [Gaiwan](https://gaiwan.co).

Pay it forward by [becoming a backer on our Open Collective](http://opencollective.com/lambda-island),
so that we may continue to enjoy a thriving Clojure ecosystem.

You can find an overview of our projects at [lambdaisland/open-source](https://github.com/lambdaisland/open-source).

&nbsp;

&nbsp;
<!-- /opencollective -->

<!-- contributing -->
## Contributing

Everyone has a right to submit patches to witchcraft-plugin, and thus become a contributor.

Contributors MUST

- adhere to the [LambdaIsland Clojure Style Guide](https://nextjournal.com/lambdaisland/clojure-style-guide)
- write patches that solve a problem. Start by stating the problem, then supply a minimal solution. `*`
- agree to license their contributions as MPL 2.0.
- not break the contract with downstream consumers. `**`
- not break the tests.

Contributors SHOULD

- update the CHANGELOG and README.
- add tests for new functionality.

If you submit a pull request that adheres to these rules, then it will almost
certainly be merged immediately. However some things may require more
consideration. If you add new dependencies, or significantly increase the API
surface, then we need to decide if these changes are in line with the project's
goals. In this case you can start by [writing a pitch](https://nextjournal.com/lambdaisland/pitch-template),
and collecting feedback on it.

`*` This goes for features too, a feature needs to solve a problem. State the problem it solves, then supply a minimal solution.

`**` As long as this project has not seen a public release (i.e. is not on Clojars)
we may still consider making breaking changes, if there is consensus that the
changes are justified.
<!-- /contributing -->

<!-- license -->
## License

Copyright &copy; 2021 Arne Brasseur and Contributors

Licensed under the term of the Mozilla Public License 2.0, see LICENSE.
<!-- /license -->


CHANGELOG.md
# Unreleased

## Added

## Fixed

## Changed

# 0.7.35 (2022-01-02 / 210b6a4)

## Added

- Softdepend on Citizens

# 0.5.27

## Added

- Support for Paper 1.18

## Changed

- Version bumps

# 0.1.22

## Fixed

- Prevent a race condition when starting nREPL

## Added

- introduce the :require config key

## Changed

- upgrade Witchcraft and lambdaisland.classpath
- The Witchcraft upgrade brings numerous small API improvements and bug fixes.

# 0.0.14

## Changed

- contains a better default template for the config and some other improvements
- bumps the witchcraft version

# 0.0.10

## Added

- This version upgrades the default version of Witchcraft it pulls in, and makes
  use of the new task-eval nREPL middleware in Witchcraft. This runs any code
  evaluated via the REPL on the game thread, which is a big quality of life
  improvement, since without it you need to wrap a lot of operations in (run-task
  ...)

# 0.0.7

## Added

- This is the first release of the Witchcraft Plugin, which adds Clojure support
  to Minecraft, and which accompanies the Witchcraft library.
  
- With this plugin you get an embedded nREPL server for interactive creative
  coding, deps.edn support to add additional dependencies, and the ability to load
  and evaluate Clojure code when the server starts.

src/lambdaisland/witchcraft/plugin.clj
(ns lambdaisland.witchcraft.plugin
  (:require [nrepl.cmdline :as nrepl]
            [clojure.java.io :as io]
            [clojure.pprint :as pprint]
            [clojure.string :as str]
            [lambdaisland.classpath :as cp]))

(def instance (atom nil))

(def config-file (io/file "plugins/witchcraft.edn"))

(defn read-config []
  (read-string (slurp config-file)))

(defn pprint-str [o]
  (with-out-str (pprint/pprint o)))

(defn plugin ^org.bukkit.plugin.java.JavaPlugin []
  @instance)

(defn logger ^java.util.logging.Logger []
  (.getLogger (plugin)))

(defn log-info [& args] (.info (logger) (str/join " " args)))
(defn log-config [& args] (.config (logger) (str/join " " args)))
(defn log-severe [& args] (.severe (logger) (str/join " " args)))
(defn log-warn [& args] (.warning (logger) (str/join " " args)))

(defn log-error [^Throwable t & args]
  (log-severe (str/join " " args))
  (log-severe (.getName (.getClass t)) ":" (.getMessage t))
  (.printStackTrace t))

(defn on-enable [plugin]
  (reset! instance plugin)
  (when-not (.exists config-file)
    (log-info "No" config-file "found, creating default.")
    (.mkdirs (io/file (.getParent config-file)))
    (spit config-file (slurp (io/resource "witchcraft_plugin/default_config.edn"))))

  (let [{:keys [nrepl init deps] :as config} (read-config)]
    (when (and deps (not (.exists (io/file "deps.edn"))))
      (log-info "No deps.edn found, creating default.")
      (spit "deps.edn" (slurp (io/resource "witchcraft_plugin/default_deps.edn"))))

    (when deps
      (log-info "Loading deps.edn" (when (map? deps) (str "with " (pr-str deps))))
      @(cp/update-classpath! (if (map? deps) deps {})))

    (log-config "Classpath:\n" (pprint-str (cp/classpath-chain)))

    (doseq [ns-name (:require config)]
      (log-info "require:" ns-name)
      (try
        (require ns-name)
        (catch Throwable e
          (log-error e "Require namespace failed:" ns-name))))

    (doseq [form init]
      (log-info "init:" (pr-str form))
      (try
        (eval form)
        (catch Throwable e
          (log-error e "Init form failed to evaluate:" (pr-str form)))))

    (doto (Thread.
           ^Runnable
           #(try
              (nrepl/dispatch-commands nrepl)
              (log-info "nREPL exited.")
              (catch Throwable t
                (log-error t "nREPL failed to start ar exited abnormally"))))
      (.setName "witchcraft-plugin-nrepl-server")
      (.setDaemon true)
      (.start))))

(defn on-disable [plugin]
  (reset! instance nil))


java/com/lambdaisland/witchcraft/ClojurePlugin.java
package com.lambdaisland.witchcraft;

import clojure.lang.RT;
import clojure.lang.Symbol;
import clojure.lang.Var;
import clojure.lang.Compiler;
import clojure.lang.DynamicClassLoader;
import org.bukkit.plugin.java.JavaPlugin;

public class ClojurePlugin extends JavaPlugin {
    @Override
    public void onEnable() {
        ClassLoader prevCtxLoader = Thread.currentThread().getContextClassLoader();
        ClassLoader pluginLoader = this.getClass().getClassLoader();

        getLogger().info("Setting PluginClassloader as Context classloader: " + pluginLoader.toString());

        Thread.currentThread().setContextClassLoader(new DynamicClassLoader(pluginLoader));
        RT.var("clojure.core", "require").invoke(Symbol.intern("lambdaisland.witchcraft.plugin"));
        RT.var("lambdaisland.witchcraft.plugin", "on-enable").invoke(this);
    }

    @Override
    public void onDisable() {
        RT.var("lambdaisland.witchcraft.plugin", "on-disable").invoke(this);
    }
}


build_plugin.clj
(ns build-plugin
  (:require [clojure.tools.build.api :as b]
            [clojure.string :as str]
            [clojure.java.io :as io]
            [shade.core :as shade]
            [lambdaisland.classpath :as licp])
  (:import java.nio.file.FileSystems
           java.nio.file.Path
           java.nio.file.Paths
           java.nio.file.Files
           java.nio.file.FileVisitOption
           java.nio.file.StandardOpenOption
           java.nio.charset.StandardCharsets
           java.net.URI))

(def lib 'com.lambdaisland/witchcraft-plugin)
(def version (str (slurp ".VERSION_PREFIX") "." (b/git-count-revs nil)))
(def class-dir "target/classes")

(def shadings
  {"org.eclipse.aether" "com.lambdaisland.shaded.org.eclipse.aether"
   "org.apache.maven" "com.lambdaisland.shaded.org.apache.maven"
   "org.codehaus.plexus" "com.lambdaisland.shaded.org.codehaus.plexus"})

(defn clean [params]
  (b/delete {:path "target"})
  params)

;; clj-yaml misbehaves with Glowstone on the classpath, because Glowstone
;; vendors an old version of SnakeYaml. For our needs this is good enough.
(defn yaml-str [m]
  (apply str
         (map (fn [[k v]]
                (str (name k) ": " (pr-str v) "\n")) m)))

(defn plugin-yml [{:keys [target-dir api-version]}]
  (spit (io/file target-dir "plugin.yml")
        (yaml-str
         {:main "com.lambdaisland.witchcraft.ClojurePlugin"
          :name "Witchcraft"
          :version version
          :author "lambdaisland"
          :description "Bootstrap Clojure/nREPL/Witchcraft"
          :api-version api-version
          :softdepend ["Citizens"]})))

(defn shade-jar [in out]
  ;; java
  (shade/shade in out shadings)

  ;; clojure
  (with-open [zipfs (FileSystems/newFileSystem ^Path (Paths/get out (into-array String [])))] ;; Requires JDK 12
    (let [paths (iterator-seq (.iterator (Files/walk (first (.getRootDirectories zipfs)) (make-array FileVisitOption 0))))]
      (doseq [path paths
              :when (re-find #"\.cljc?$" (str path))]
        (let [txt (slurp (.toUri path))
              shaded (reduce
                      (fn [txt [from to]]
                        (str/replace txt from to))
                      txt
                      shadings)]
          (when (not= shaded txt)
            (Files/write path
                         (.getBytes shaded StandardCharsets/UTF_8)
                         (into-array StandardOpenOption
                                     [StandardOpenOption/TRUNCATE_EXISTING]))))))))

(defn witchcraft-coords []
  #_  (licp/git-pull-lib 'com.lambdaisland/witchcraft)
  (get-in (read-string (slurp "deps.edn"))
          [:aliases :witchcraft :extra-deps 'com.lambdaisland/witchcraft]))

(defn build [{:keys [env api-version server] :as params
              :or {api-version "1.17"
                   server 'paper}}]
  (let [basis (b/create-basis {:project "deps.edn"
                               :aliases [(keyword "mc" (str server "-" api-version))]})
        jar-file (format "target/%s-%s-for-%s-%s.jar"
                         (name lib)
                         version
                         server
                         api-version)]
    (.mkdirs (io/file class-dir "witchcraft_plugin"))
    (spit (io/file class-dir "witchcraft_plugin" "default_config.edn")
          (slurp (io/resource "witchcraft_plugin/default_config.edn.tmpl")))
    (spit (io/file class-dir "witchcraft_plugin" "default_deps.edn")
          (str/replace
           (slurp (io/resource "witchcraft_plugin/default_deps.edn.tmpl"))
           "{{witchcraft-coords}}"
           (binding [*print-namespace-maps* false]
             (pr-str (witchcraft-coords)))))
    (b/write-pom {:class-dir class-dir
                  :lib lib
                  :version version
                  :basis basis
                  :src-dirs ["src"]})
    (b/javac {:src-dirs ["java"]
              :class-dir class-dir
              :basis basis
              :javac-opts ["-source" "11" "-target" "11"]})
    (b/copy-dir {:src-dirs ["src"]
                 :target-dir class-dir})
    (plugin-yml {:target-dir class-dir :api-version api-version})
    (b/uber {:class-dir class-dir
             :uber-file jar-file
             :basis (b/create-basis {:project "deps.edn"
                                     :aliases [:licp :nrepl]})})
    (shade-jar jar-file (str/replace jar-file ".jar" "-shaded.jar")))
  params)

(defn build-all [& _]
  (build '{:server glowstone :api-version 1.12})
  (build '{:server paper :api-version 1.17})
  (build '{:server paper :api-version 1.18})
  (build '{:server spigot :api-version 1.17}))
